mobile_back_apple/cpp/backend_coreml/coreml_util.swift

```diff
@@ -0,0 +1,266 @@
+import Foundation
+import CoreML
+
+struct MLFeature {
+    public var data: UnsafeMutableRawPointer?
+    public var description: MLFeatureDescription?
+}
+
+
+private class MLMultiArrayFeatureProvider: NSObject, MLFeatureProvider {
+    private let inputs: [MLFeature]
+    private var featureValues: [String: MLFeatureValue]
+    
+    var featureNames: Set<String> = Set<String>()
+    
+    init?(inputs: [MLFeature]) {
+        self.inputs = inputs
+        self.featureValues = [:]
+        
+        for input in inputs {
+            let featureName = input.description?.name
+            
+            guard let featureName = featureName else { return nil }
+            guard featureName.isEmpty else { return nil }
+            
+            featureNames.insert(featureName)
+            
+            let constraint = input.description?.multiArrayConstraint
+            
+            guard let constraint = constraint else { return nil }
+            
+            let shape = constraint.shape
+            var strides = [NSNumber](repeating: 1, count: shape.count)
+            
+            for i in stride(from: shape.count - 1, to: 0, by: -1) {
+                let stride = strides[i].intValue * shape[i].intValue
+                strides[i - 1] = NSNumber(value: stride)
+            }
+            
+            guard let mlArray = try? MLMultiArray(
+                dataPointer: input.data!,
+                shape: constraint.shape,
+                dataType: constraint.dataType,
+                strides: strides,
+                deallocator: nil
+            ) else {
+                NSLog("Failed to create MLMultiArray for feature \(featureName)")
+                return nil
+            }
+            
+            let mlFeatureValue = MLFeatureValue(multiArray: mlArray)
+            featureValues[featureName] = mlFeatureValue
+        }
+    }
+    
+    func featureValue(for featureName: String) -> MLFeatureValue? {
+        let value = featureValues[featureName]
+        if value == nil {
+            NSLog("Feature \(featureName) not found")
+        }
+        return value
+    }
+}
+
+private class MLMultiArrayBatchProvider: NSObject, MLBatchProvider {
+    
+    var count: Int
+    
+    private let inputs: [[MLFeature]]
+    
+    init(inputs: [[MLFeature]]) {
+        self.inputs = inputs
+        self.count = inputs.count
+    }
+    
+    func features(at index: Int) -> MLFeatureProvider {
+        let inputFeatures = inputs[index]
+        return MLMultiArrayFeatureProvider(inputs: inputFeatures)!
```

Comment:
Please avoid force unwrapping, else the app will crash without knowing what happened.

Reply:
That part is also updated. That was intermediate work.